### PR TITLE
294 build state machine for application flow

### PR DESCRIPTION
**Pull request by afogel**
## What's the change?
This introduces the state machine to handle transitions through the mentee application flow, based on the flow laid out in the [notion document](https://www.notion.so/toyhammered/Application-Submission-Design-Doc-7a45fd8e25ce4f448e9b47bc00a142b5).

The existing ServiceObject conforms to the API laid out in the [notion design doc](https://www.notion.so/toyhammered/Applicant-State-Machine-Design-Doc-eb61b1dc2ae4417da53ada037cea3916) that I published yesterday.

## What key workflows are impacted?
Currently, there are none. In the next PR I upload, I will hook it up to the existing forms/controllers

## Highlights / Surprises / Risks / Cleanup

1) we added an addition state the enum for MenteeApplicationState, so I've included a migration so that the change doesn't impact the mapping for existing applications that have already gone through the funnel
2) @dpaola2 @toyhammered I specifically want to call attention to the copy of the new mailers that will get sent off for acceptances/rejections. 

For rejections, I used the more verbose copy I drafted since it reflects the YC-style reapplication ethos while using a tone of voice more reflective of AOL's culture.

For Acceptances, I wasn't clear on the appropriate CTAs that we want to communicate after we notify applicants that they were accepted. Please let me know what I can tweak

3) I assume that we probably want the state machine to handle sending out the notification email with the code challenge, but for now I've omitted it since it was not clear. If that falls within scope, I will update to also send a notification when promoted to `:code_challenge_sent`.

## Issue ticket number and link
#294 

## Checklist before requesting a review

Please delete items that are not relevant.

- [X] Did you add appropriate automated tests?
- [X] Did you consider risks around security, performance, etc.?
- [X] Have you thought of misfiring code? e.g. too many loops, n+1, or how to handle nils?
- [X] Any validations to data needed?
- [X] Related to readability and consistency: Did you add comments and/or documentation? (README, Notion, etc.)
- [X] Is everything elevated to the highest level? (e.g., can logic be moved out of view into controllers, or out of controllers into models?)
- [X] Can (and should) any models be extracted?
- [X] Is there any linting that needs to be executed?
- [X] Did you leave helpful inline PR comments for the reviewer(s)?
- [X] Do you have a plan for how this change should be rolled back? (e.g., how do you deal with a migration rollback? Is your feature released behind a feature flag?)

### CODE REVIEW

File: spec/models/mentee_application_state_spec.rb [hunk 1]
```diff
@@ -24,40 +24,48 @@
 
 RSpec.describe MenteeApplicationState do
   describe '.next' do
-    context 'when the status is pending' do
-      let(:status) { :pending }
+    context 'when the status is application_received' do
+      let(:status) { :application_received }
```
Comment (afogel) [inline spec/models/mentee_application_state_spec.rb:28]: while I plan on ripping out these specs in the next PR, I want to make sure to update them so that the existing functionality is not changed.

File: spec/models/mentee_application_state_spec.rb [hunk 2]
```diff
@@ -24,40 +24,48 @@
...
 
       it 'returns the next status' do
-        expect(described_class.next(status:)).to eq('stage_four')
+        expect(described_class.next(status:)).to eq('phone_screen_scheduled')
       end
     end
 
-    context 'when the status is stage_four' do
-      let(:status) { :stage_four }
+    context 'when the status is phone_screen_scheduled' do
```
Comment (afogel) [inline spec/models/mentee_application_state_spec.rb:59]: new test reflects added state that was added in updated flow.

File: db/migrate/20230912072042_update_mentee_application_state_stages.rb [hunk 3]
```diff
@@ -0,0 +1,17 @@
+class UpdateMenteeApplicationStateStages < ActiveRecord::Migration[7.0]
+  def up
+    rejected_applications = MenteeApplicationState.where(status: 6)
```
Comment (afogel) [inline db/migrate/20230912072042_update_mentee_application_state_stages.rb:3]: migration to map `:accepted` and `:rejected` to the correct value before casting.

File: app/notifications/mentee_application/rejection_notification.rb [hunk 4]
```diff
@@ -0,0 +1,25 @@
+# To deliver this notification:
+#
+# MenteeApplication::RejectionNotification.with(post: @post).deliver_later(current_user)
+# MenteeApplication::RejectionNotification.with(post: @post).deliver(current_user)
+
+class MenteeApplication::RejectionNotification < Noticed::Base
+  deliver_by :email,
+    mailer: 'MenteeApplicationMailer',
+    method: :notify_for_rejection,
+    enqueue: true
```
Comment (afogel) [inline app/notifications/mentee_application/rejection_notification.rb:10]: TIL: `enqueue: true` is necessary for email notifications to be enqueued as background jobs. Without this, the tests that checked that there was a Mailer enqueued would fail. I think being explicit here is helpful when TDDing and making sure that the mailer work is performed asynchronously.

File: config/environments/test.rb [hunk 5]
```diff
@@ -59,4 +59,5 @@
   # config.action_view.annotate_rendered_view_with_filenames = true
   config.action_mailer.perform_deliveries = false
   config.action_mailer.default_url_options = { host: 'localhost', port: 3000 }
+  config.active_job.queue_adapter = :test
```
Comment (afogel) [inline config/environments/test.rb:62]: necessary to make the assertions like:
```ruby
expect {
  described_class.promote!(application: phone_screen_completed, user: reviewer)
}.to have_enqueued_mail(MenteeApplicationMailer, :notify_for_acceptance)
```

File: app/services/mentee_application_transition_service.rb [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+  end
+
+  def reject!(application:, user:, note: nil)
+    status = application.current_status.to_sym
+    raise InvalidTransitionError unless STATUS_TRANSITION_MAPPING[status][:valid_transitions].include? :reject
+    promote_transition = :rejected
+    application.mentee_application_states.create!(status: promote_transition, status_changed: user, note:)
+    invoke_side_effects(application, promote_transition, user)
+    true
+  end
```
Comment (toyhammered) [inline app/services/mentee_application_transition_service.rb:54]: A more open question of do we want to keep this stateless like we have now or allow for `application` and `user` to be passed in as instances (and i'll call it semi stateless) but give access to nice instance variables? I am fine either way (I do have a preference but not a big deal)

File: app/services/mentee_application_transition_service.rb [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+    },
+    accepted: {
+      valid_transitions: []
+    },
+    rejected: {
+      valid_transitions: []
+    }
+  }.freeze
+
+  def promote!(application:, user:, note: nil)
```
Comment (toyhammered) [inline app/services/mentee_application_transition_service.rb:38]: We should rename this `user` to something more clear (I got confused it was the person who is accepting/rejecting). Idk of a good name for them but it shouldn't be just user.

File: db/migrate/20230912072042_update_mentee_application_state_stages.rb (hunk 3 above)
Comment (toyhammered) [reply db/migrate/20230912072042_update_mentee_application_state_stages.rb:3]: oh nice! totally missed this.

**Review by toyhammered — COMMENTED** [review]
This looks great! Left a few comments/questions. 

Also to answer your question about the `code_challenge_sent`, yes make sure that is included here.

File: app/services/mentee_application_transition_service.rb (hunk 6 above)
Comment (afogel) [reply app/services/mentee_application_transition_service.rb]: What do you consider the larger benefits being of instantiating the service with application and reviewer? Though I considered instantiating, it seemed more reasonable to reach for a PORO singleton type thing and have the service act as I/O as possible. 

tl;dr My preference is stateless -- I think the state that this service should have is only the state machine transitions, but would be open to better understanding why you might want to pass in the application and user.

Comment (toyhammered) [reply app/services/mentee_application_transition_service.rb]: > 

oh totally fine with that. Was more curious as to your thoughts. I am totally fine with this. I sometimes prefer instance variables so we can use them in other methods more easily (typically I call it like stateless instantiation, because the service should only be run once).

**Review by toyhammered — APPROVED** [review]
One small thing, but we should also wait for Dave to review (I am also open to him reviewing later on and we can open up another PR to fix stuff).

File: app/views/mentee_application_mailer/notify_for_acceptance.html.erb [hunk 8]
```diff
@@ -0,0 +1,11 @@
+
+<div style='display: block; max-width: 36rem; margin: 0 auto; padding-top: 4rem;'>
+  <p><%= @recipient.full_name %>,</p>
+  <p>
+    We are pleased to welcome you to the <strong>Agency of Learning</strong>! 🎉🎉🎉
+  </p>
+
+  <p>
+    Your cohort begins on <%= @application.user_mentee_application_cohort&.begin&.to_fs(:long) %>. Please send an email to <a href="mailto:dave@agencyoflearning.com?subject=Accept Invitation to Join Agency of Learning">dave@agencyoflearning.com</a> confirming your interest to join us in the program. We're excited to start working with you!
```
[the remark this answers is not in the document]
Comment (dpaola2) [reply app/views/mentee_application_mailer/notify_for_acceptance.html.erb:9]: why are we instructing them to email me after they've been accepted?

File: app/views/mentee_application_mailer/notify_for_rejection.html.erb [hunk 9]
```diff
@@ -0,0 +1,16 @@
+
+<div style='display: block; max-width: 36rem; margin: 0 auto; padding-top: 4rem;'>
+  <p><%= @recipient.full_name %>,</p>
+
+  <p>
+    Thank you for submitting your application to join us at the <strong>Agency of Learning</strong>!
+  </p>
+
+  <p>
+    Unfortunately, we got more applicants than we could admit in this current cohort. Though we know this can feel discouraging, we strongly encourage you to <strong>reapply for the next cohort</strong>. We take into consideration the progress you've made since your last application.
```
Comment (dpaola2) [inline app/views/mentee_application_mailer/notify_for_rejection.html.erb:10]: I have already copy edited this somewhere, this is old copy

Comment (dpaola2) [reply app/views/mentee_application_mailer/notify_for_rejection.html.erb:10]: we do not want to tell them how to feel about this, let alone instruct them that they should feel discouraged!

Comment (dpaola2) [reply app/views/mentee_application_mailer/notify_for_rejection.html.erb:10]: Thank you for submitting your application to join us at the Agency of Learning!

Unfortunately you have not been accepted into the upcoming batch. This was a tough decision. We can't make any promises about being admitted on a future application, but we hope that you reapply for a future batch.

File: app/views/mentee_application_mailer/notify_for_rejection.html.erb [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+  </p>
+
+  <p>
+    We'll be candid -- though we aspire to have many future cohorts, currently we are only able to support smaller cohorts. Even though we can't make any promises about being admitted on a future application, we hope that you reapply and are excited to help you as you transition into the next stage of your career.
```
Comment (dpaola2) [inline app/views/mentee_application_mailer/notify_for_rejection.html.erb:14]: get rid of this

File: app/views/mentee_application_mailer/notify_for_acceptance.html.erb (hunk 8 above)
[the remark this answers is not in the document]
Comment (toyhammered) [reply app/views/mentee_application_mailer/notify_for_acceptance.html.erb:9]: > why are we instructing them to email me after they've been accepted?

I also have a semi question about it. I think the better approach would be to just update them to member and tell them to join in now.

[the remark this answers is not in the document]
Comment (afogel) [reply app/views/mentee_application_mailer/notify_for_acceptance.html.erb:9]: yep and yep. I wasn't sure whether they're they're just in or what other steps you might want to include. That was part of the reason I wanted to run this by you. I'll just send an invitation to the discord.

File: app/views/mentee_application_mailer/notify_for_rejection.html.erb (hunk 9 above)
Comment (afogel) [reply app/views/mentee_application_mailer/notify_for_rejection.html.erb:10]: sounds good, thanks for the help with the copy.

**Comment by afogel** [comment]
@toyhammered merging since the changes I made since last review are small, but I wanted to call out the changes so that they're on your radar in case we want to address them in a future PR:
1. copy changes to rejection email to reflect what Dave wrote. I appreciate the consistency, and I think Dave's call-out to not tell applicants how to feel is an important reminder to me as I improve in my copy-writing.
2. changes to the acceptance email -- I'm not directing them to join the discord channel via a non-expiring link (as a first step, we can always revise).
3. I'm updating the applicant's role to `member` as an added side effect of their application getting promoted to `accepted` (and tests).
